Implement pending specs for StatusesController

mjankowski opened the pull request:
Updates a collection of specs for the `activity` action which were written and outlined but not implemented.

I have a question here before this is merged ... for the specs where the user is signed in and we have either a `private` or `direct` visibility status -- what are the actual conditions which lead to the user being authorized to see it or not, in the context of this controller action?

From a quick look at the code, there's a check for `status.distributable?` in place, but that in turn is just checking for public or unlisted visbility -- so it seems like we'd always return 404 when not in those states, regardless of which user was making the request.

It strikes me as possible that these specs were outlined and the behavior has changed since then so the descriptions are no longer relevant ... or that I'm missing something in how the code processes these different types of requests.

After I get clarity on this I'll edit this PR to update as relevant, and also add in the specs for the "signature" block in this spec which is still pending.

mjankowski committed:
Implement pending specs for StatusesController

Updates a collection of specs for the `activity` action which were
written and outlined but not implemented.

nschonni commented:
@ClearlyClaire is this PR gets resolved, it would get rid of most of the noise from the GitHub annotations reporter in https://github.com/mastodon/mastodon/pull/24040

ClearlyClaire commented:
Sorry it took me the whole week to get back to you!

`activity` is useful through ActivityPub only, although the way it is implemented allows logged-in local users to access them too.

In both cases, private and DMs are only accessible from people with the correct rights. The `status.distributable?` check is for whether it should be cached.

mjankowski commented:
Ok, got it ... will attempt to update this to find scenarios where the signed in user is able to see the statuses for the private and direct cases; and then apply the whole thing to the auth-with-signature block as well.

mjankowski commented:
Updated so that the `direct` and `private` visibilities for signed in users get handled correctly, and also added matching specs for the authing with signature cases. Ready for review.
